mypy/subtypes.py

```diff
@@ -302,6 +331,237 @@ def visit_type_type(self, left: TypeType) -> bool:
+    we pop the assumption, thus avoiding an infinite recursion as well.
+    """
+    assert right.type.is_protocol
+    assuming = right.type.assuming if allow_any else right.type.assuming_proper
+    for (l, r) in reversed(assuming):
+        if sametypes.is_same_type(l, left) and sametypes.is_same_type(r, right):
+            return True
+    with pop_on_exit(assuming, left, right):
+        for member in right.type.protocol_members:
+            # nominal subtyping currently ignores '__init__' and '__new__' signatures
+            if member in ('__init__', '__new__'):
+                continue
+            # The third argiment below indicates to what self type is bound.
+            # We always bind self to the subtype. (Similarly to nominal types).
+            supertype = find_member(member, right, left)
+            assert supertype is not None
+            subtype = find_member(member, left, left)
+            # Useful for debugging:
+            # print(member, 'of', left, 'has type', subtype)
+            # print(member, 'of', right, 'has type', supertype)
+            if not subtype:
+                return False
+            if allow_any:
+                # nominal check currently ignores arg names
+                is_compat = is_subtype(subtype, supertype, ignore_pos_arg_names=True)
+            else:
+                is_compat = is_proper_subtype(subtype, supertype)
+            if not is_compat:
+                return False
+            if isinstance(subtype, NoneTyp) and member.startswith('__') and member.endswith('__'):
+                # We want __hash__ = None idiom to work even without --strict-optional
+                return False
+            subflags = get_member_flags(member, left.type)
+            superflags = get_member_flags(member, right.type)
+            if IS_SETTABLE in superflags:
+                # Check opposite direction for settable attributes.
+                if not is_subtype(supertype, subtype):
+                    return False
+            if (IS_CLASSVAR in subflags and IS_CLASSVAR not in superflags or
+                    IS_CLASSVAR in superflags and IS_CLASSVAR not in subflags):
+                return False
+            if IS_SETTABLE in superflags and IS_SETTABLE not in subflags:
+                return False
+            # this rule is copied from nominal check in checker.py
+            if IS_CLASS_OR_STATIC in superflags and IS_CLASS_OR_STATIC not in subflags:
+                return False
+    if allow_any:
+        right.type.cache.add((left, right))
+    else:
+        right.type.cache_proper.add((left, right))
+    return True
+
+
+def find_member(name: str, itype: Instance, subtype: Type) -> Optional[Type]:
+    """Find the type of member by 'name' in 'itype's TypeInfo. Apply type arguments
+    from 'itype', and bind 'self' to 'subtype'. Return None if member was not found.
+    """
+    info = itype.type
```

Comment:
This seems to duplicate a bunch of logic from `checkmember.py`. At least at a TODO comment about combining these (to both locations), as otherwise it's likely that these will soon get out of sync.